pkg/kubernetes/deployment/v1alpha1/kubernetes.go

```diff
@@ -0,0 +1,199 @@
+			if err != nil {
+				return nil, err
+			}
+			return kubernetes.NewForConfig(config)
+		}
+	}
+
+	if k.getExtnV1Beta1 == nil {
+		k.getExtnV1Beta1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (d *api_extn_v1beta1.Deployment, err error) {
+			d = &api_extn_v1beta1.Deployment{}
+			err = cli.ExtensionsV1beta1().
+				RESTClient().
+				Get().
+				Namespace(namespace).
+				Name(name).
+				Resource("deployments").
+				Do().
+				Into(d)
+			return
+		}
+	}
+
+	if k.getAppsV1 == nil {
+		k.getAppsV1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (d *api_apps_v1.Deployment, err error) {
+			d = &api_apps_v1.Deployment{}
+			err = cli.AppsV1().
+				RESTClient().
+				Get().
+				Namespace(namespace).
+				Name(name).
+				Resource("deployments").
+				Do().
+				Into(d)
+			return
+		}
+	}
+
+	if k.rollOutStatusExtnV1Beta1 == nil {
+		k.rollOutStatusExtnV1Beta1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (op []byte, err error) {
+			d, err := k.getExtnV1Beta1(cli, name, namespace)
+			if err != nil {
+				return nil, err
+			}
+			return DeployExtnV1Beta1(WithExtnV1Beta1Deployment(d)).
+				AddCheck(IsSyncSpecV1B1()).
+				AddCheck(IsProgressDeadlineExceededV1B1()).
+				AddCheck(IsTerminationInProgressV1B1()).
+				AddCheck(IsUpdationInProgressV1B1()).
+				AddCheck(IsOlderReplicaActiveV1B1()).
+				RollOutStatus()
+		}
+	}
+
+	if k.rollOutStatusAppsV1 == nil {
+		k.rollOutStatusAppsV1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (op []byte, err error) {
+			d, err := k.getAppsV1(cli, name, namespace)
+			if err != nil {
+				return nil, err
+			}
+			return DeployAppsv1(WithAppsv1Deployment(d)).
+				AddCheck(IsSyncSpecV1()).
+				AddCheck(IsProgressDeadlineExceededV1()).
+				AddCheck(IsTerminationInProgressV1()).
+				AddCheck(IsUpdationInProgressV1()).
+				AddCheck(IsOlderReplicaActiveV1()).
+				RollOutStatus()
+		}
+	}
+
+}
+
+// WithKubeClient sets the kubernetes client against the kubeclient instance
+func WithKubeClient(c *kubernetes.Clientset) kubeclientBuildOption {
+	return func(k *kubeclient) {
+		k.clientset = c
+	}
+}
+
+// KubeClient returns a new instance of kubeclient meant for deployment.
+// caller can configure it with different kubeclientBuildOption
+func KubeClient(opts ...kubeclientBuildOption) *kubeclient {
+	k := &kubeclient{}
+	for _, o := range opts {
+		o(k)
+	}
+	k.withDefaults()
+	return k
+}
+
+// getClientOrCached returns either a new instance of kubernetes client or its cached copy
+func (k *kubeclient) getClientOrCached() (*kubernetes.Clientset, error) {
+	if k.clientset != nil {
+		return k.clientset, nil
+	}
+	c, err := k.getClientset()
+	if err != nil {
+		return nil, err
+	}
+	k.clientset = c
+	return k.clientset, nil
+}
+
+// GetExtnV1Beta1 returns deployment(ExtnV1Beta1) object for given name and namespaces
+func (k *kubeclient) GetExtnV1Beta1(name, namespace string) (*api_extn_v1beta1.Deployment, error) {
```

Comment:
By separating deployment logic into different packages, this function can be more generic.
Later when we have interface we can expose these functions as-is.

Refer to https://github.com/openebs/maya/blob/master/pkg/openebscluster/v1alpha1/interface.go
Above is just FYI only.
Do not implement above interface.